contracts/cw1-subkeys/src/contract.rs

```diff
@@ -436,6 +438,138 @@ mod tests {
+        }
+    }
+
+    #[derive(Default)]
+    struct Spender {
+        spender: &'static str,
+        allowances: Vec<Coin>,
+        allowances_expire: Option<Expiration>,
+        permissions: Option<Permissions>,
+    }
+
+    /// Test suite helper unifying test initialization, keeping access to created data
+    struct Suite {
+        deps: OwnedDeps<MockStorage, MockApi, MockQuerier>,
+        owner_info: MessageInfo,
+    }
+
+    impl Suite {
+        /// Initializes test case using default config
+        fn init() -> Self {
+            Self::init_with_config(SuiteConfig::default())
+        }
+
+        /// Initialized test case using provided config
+        fn init_with_config(config: SuiteConfig) -> Self {
+            let mut deps = mock_dependencies(&[]);
+            let admins = std::iter::once(OWNER)
+                .chain(config.admins)
+                .map(ToOwned::to_owned)
+                .collect();
+
+            let instantiate_msg = InstantiateMsg {
+                admins,
+                mutable: true,
+            };
+            let owner_info = mock_info(OWNER, &[]);
+
+            instantiate(
+                deps.as_mut().branch(),
+                mock_env(),
+                owner_info.clone(),
+                instantiate_msg,
+            )
+            .unwrap();
+
+            for Spender {
+                spender,
+                allowances,
+                allowances_expire: expires,
+                permissions,
+            } in config.spenders
+            {
+                for amount in allowances {
+                    let msg = ExecuteMsg::IncreaseAllowance {
+                        spender: spender.to_owned(),
+                        amount,
+                        expires,
+                    };
+                    execute(deps.as_mut().branch(), mock_env(), owner_info.clone(), msg).unwrap();
+                }
+
+                if let Some(permissions) = permissions {
+                    let msg = ExecuteMsg::SetPermissions {
+                        spender: spender.to_owned(),
+                        permissions,
+                    };
+                    execute(deps.as_mut().branch(), mock_env(), owner_info.clone(), msg).unwrap();
+                }
+            }
+
+            Self { deps, owner_info }
+        }
+    }
+
+    /// Helper function for comparing vectors or another slice-like object as they would represent
+    /// set with duplications. Compares sets by first sorting elements usning provided ordering.
+    /// This functions reshufless elements inplace, as it should never matter as compared
+    /// containers should represent same value regardless of ordering, and making this inplace just
+    /// safes obsolete copying.
+    ///
+    /// This is implemented as a macro instead of function to throw panic in the place of macro
+    /// usage instead of from function called inside test.
+    macro_rules! assert_sorted_eq {
```

Comment:
Interesting macro trick.

I write a number of helper functions and always re-run with `RUST_BACKTRACE=1` when they fail to find the real location. This does seem more ergonomic.

We need to reconsider the macro-resistance in this codebase.